### DISCUSSION TITLE
How can I call self method from from `Command::perform`?

### DISCUSSION THREAD

**Discussion by H36615** [hecrj/iced#1012] [Q&A] [answered]
```
     fn update(&mut self, message: Message) -> Command<Message> {
        match message {
            Message::DataRequested => {
                Command::perform(self.controller.get_data_async(), Message::DataResponded)
            }
            ...
        }
    }
```

```
error[E0759]: `self` has an anonymous lifetime `'_` but it needs to satisfy a `'static` lifetime requirement
  --> ...\lib.rs:57:50
   |
54 |     fn update(&mut self, message: Message) -> Command<Message> {
   |               --------- this data with an anonymous lifetime `'_`...
...
57 |                 Command::perform(self.controller.get_data_async(), Message::DataResponded)    
   |                                  --------------- ^^^^^^^^^
   |                                  |
   |                                  ...is captured here...
   |
note: ...and is required to live as long as `'static` here
  --> ...\lib.rs:57:17
   |
57 |                 Command::perform(self.controller.get_data_async(), Message::DataResponded) 
   |                 ^^^^^^^^^^^^^^^^
```

Is it possible to call `self.prop.method` from `Command::perform` even though it's requiring self to have static lifetime? I hope the answer is not just `Controller::get_data_async()`.

TIA :)

**Comment by 90-008** [comment, marked as answer]
Your `get_data_async` method borrows some data from `self`, and `Command`s require `'static` lifetime, but `self` is not `'static`, so Rust complains. You should `Clone` (`.clone()`) whatever data you need to perform the async code in `get_data_async`, so that nothing is borrowed from `self`.

**Reply by H36615** [reply]
Thanks I'll check that out asap when I have time.

**Reply by H36615** [reply]
Sorry I don't quite follow. You mean I should clone e.g. `self.controller` or something inside `get_data_async`? What could the line 57 look like for example?

**Reply by 90-008** [reply]
I'm not sure how your `get_async_data` function looks, but you have some data that is borrowed, some stuff you borrow from `controller`, so you can try cloning them. The `get_data_async` function would then look something like this:
```rust
fn get_data_async(&self) -> impl std::future::Future<Output = output> + 'static {
  // clone data you will use in your logic
  async move {
    // your logic
  }
}
```
